lib/next_ls/commands/from_pipe.ex

```diff
@@ -0,0 +1,116 @@
+defmodule NextLS.Commands.FromPipe do
+  @moduledoc false
+  alias GenLSP.Enumerations.ErrorCodes
+  alias GenLSP.Structures.Position
+  alias GenLSP.Structures.Range
+  alias GenLSP.Structures.TextEdit
+  alias GenLSP.Structures.WorkspaceEdit
+
+  defp opts do
+    Schematic.map(%{
+      position: Schematic.map(%{character: Schematic.int(), line: Schematic.int()}),
+      uri: Schematic.str(),
+      text: Schematic.list(Schematic.str())
+    })
+  end
+
+  def new(opts) do
+    {:ok, %{text: text, uri: uri, position: position}} = Schematic.unify(opts(), Map.new(opts))
+
+    case from_pipe_edit(text, position) do
+      {:ok, %TextEdit{} = edit} ->
+        %WorkspaceEdit{
+          changes: %{
+            uri => [edit]
+          }
+        }
+
+      {:error, message} ->
+        %GenLSP.ErrorResponse{code: ErrorCodes.parse_error(), message: inspect(message)}
+    end
+  end
+
+  defp find_pipe_lines(text, line) do
+    coursor_text_line = Enum.at(text, line)
+
+    if String.contains?(coursor_text_line, "|>") do
+      case Code.string_to_quoted(coursor_text_line) do
+        {:ok, _} ->
+          range = %Range{
+            start: %Position{line: line, character: 0},
+            end: %Position{line: line, character: String.length(coursor_text_line)}
+          }
+
+          {:ok, [coursor_text_line], range}
+
+        {:error, _} ->
+          previous_line = Enum.at(text, line - 1)
+
+          range = %Range{
+            start: %Position{line: line - 1, character: 0},
+            end: %Position{line: line, character: String.length(coursor_text_line)}
+          }
+
+          {:ok, [previous_line, coursor_text_line], range}
+      end
+    else
+      next_line = Enum.at(text, line + 1)
+
+      range = %Range{
+        start: %Position{line: line, character: 0},
+        end: %Position{line: line + 1, character: String.length(next_line)}
+      }
+
+      {:ok, [coursor_text_line, next_line], range}
+    end
+  end
+
+  defp from_pipe_edit(text, %{line: line}) do
+    with {:ok, lines, range_to_edit} <- find_pipe_lines(text, line),
+         {:ok, indent} <- get_indent(text, line),
+         {:ok, edit} <- get_edit(lines) do
+      {:ok, %TextEdit{new_text: indent <> edit, range: range_to_edit}}
+    else
+      {:error, _message} = error ->
+        error
```

Comment:
I'd get rid of this, just let it blow up at the call site if any of the return values don't match